### PR TITLE
UI Automation in Windows Console: Make UIA the default for Windows 10 1803 and later

### LINKED ISSUE #513: Some output in command consoels spelled with speak typed characters enabled
**Reported by aleksey_s on 2009-12-17 16:39**
Sometimes, NVDA is too verbose in console windows.
STR:
- open cmd
- run python interpreter
- type: print "hi"
  Currently, NVDA says
  h
  i
  hi
  It starts reading by letter and only then reads the entire message. This behaviour also exists in other console apps.
  **Blocked by #5027**
  **Blocking #2977, #4127, #5328**

### LINKED ISSUE #673: Problems with command consoles and object navigation
**Reported by jteh on 2010-05-26 22:41**
When you move to a command console with object navigation, NVDA connects to this console, which allows this console to be reviewed. While this is useful, it presents three problems:
1. When you move to the console using object navigation, the console is connected, but it isn't disconnected when you move away.
   - This is because there is a becomeNavigatorObject event, but no event for the reverse.
2. When object navigation causes NVDA to connect to a console, new text is reported for that console. New text should only be reported for focused consoles.
3. Because only one console can be connected at a time, if a console is focused and the user moves to another console with object navigation, the user can no longer access the focused console. That is, new text is not read nor is caret navigation, backspace, etc.

While issues 1 and 2 can be fixed, it is probably not possible to solve issue 3, as this is due to a limitation of the implementation in Windows. Therefore, the only option may be to not connect to consoles for review.

Work around: To restore functionality to a focused console, move the navigator object back to the focused console.

This functionality was introduced in 4808ab654c473343bac0a49b185128d8e31ce116.

### PR DESCRIPTION
**Pull request by codeofdusk**
### Link to issue number:
Builds on #9614. Closes #673 and #6291.

### Summary of the issue:
NVDA's UIA console support significantly improves performance and stability, so we should enable it by default.

### Description of how this pull request fixes the issue:
While UIA in consoles was added in Windows 10 1709, the initial implementation left much to be desired, particularly concerning caret movement and expansion to character.

A new `UIAUtils.shouldUseUIAConsole` function has been added which determines whether to use UIA in consoles:
* If the user has explicitly chosen to use UIA or legacy console support, the function returns `True` or `False` respectively.
* Otherwise, if the user is running at least Windows 10 1803, the function returns `True`, or `False` if not.

The "use UI Automation in the Windows Console when available" GUI setting has been changed to expose all options in the config spec (i.e. to automatically detect UIA/legacy or override this autodetection).

### Testing performed:
Tested the new GUI setting on Windows 10 1903.

### Known issues with pull request:
None.

### Change log entry:
== New Features ==
- Vastly improved performance and stability in Command Prompt, PowerShell, and the Windows Subsystem for Linux on Windows 10 version 1803 and later. (#9771)
- In Command Prompt, PowerShell, and the Windows Subsystem for Linux on Windows 10 version 1803 and later, the reporting of typed text that does not appear onscreen (such as passwords) can now be toggled in NVDA's advanced preferences. (#9649)

### CODE REVIEW

**Comment by codeofdusk** [comment]
CC @feerrenrut @leonardder @michaelDCurran

**Comment by josephsl** [comment]
Hi, if you can, can you test this in 20H1 builds too? I myself will do some testing in 20H1 via Windows Sandbox. Thanks.

File: source/UIAUtils.py [hunk 1]
```diff
@@ -203,3 +205,11 @@ def getValue(self,ID,ignoreMixedValues=False):
 			raise UIAMixedAttributeError
 		return val
 
+def shouldUseUIAConsole():
+	"Determines whether to use UIA in the Windows Console."
```
Comment (leonardder) [inline source/UIAUtils.py]: Please wrap this in tripplets

```suggestion
	"""Determines whether to use UIA in the Windows Console."""
```

File: source/UIAUtils.py [hunk 2]
```diff
@@ -203,3 +205,11 @@ def getValue(self,ID,ignoreMixedValues=False):
...
 		return val
 
+def shouldUseUIAConsole():
+	"Determines whether to use UIA in the Windows Console."
+	setting = config.conf['UIA']['winConsoleImplementation']
+	if setting == "legacy":
+		return False
+	elif setting == "UIA":
+		return True
+	return isWin10(1809) and not isWin10(1903, atLeast=False)
```
Comment (leonardder) [inline source/UIAUtils.py]: I think I prefer this to be False for windows 10 1903 and above, until we can make sure that only 1903 will be affected by this.

File: source/_UIAHandler.py [hunk 3]
```diff
@@ -355,7 +355,10 @@ def _isBadUIAWindowClassName(self, windowClass):
 		"Given a windowClassName, returns True if this is a known problematic UIA implementation."
 		# #7497: Windows 10 Fall Creators Update has an incomplete UIA implementation for console windows, therefore for now we should ignore it.
 		# It does not implement caret/selection, and probably has no new text events.
```
Comment (leonardder) [inline source/_UIAHandler.py]: I think you can remove these comments here and move this information to the doc string of `shouldUseUIAConsole`.

File: source/gui/settingsDialogs.py [hunk 4]
```diff
@@ -2050,13 +2050,36 @@ def __init__(self, parent):
...
-		#  Advanced settings panel.
-		label = _("Use UI Automation to access the Windows Console when available")
-		consoleUIADevMap = True if config.conf['UIA']['winConsoleImplementation'] == 'UIA' else False
-		self.ConsoleUIACheckBox=UIAGroup.addItem(wx.CheckBox(self, label=label))
-		self.ConsoleUIACheckBox.SetValue(consoleUIADevMap)
-		self.ConsoleUIACheckBox.defaultValue = self._getDefaultValue(["UIA", "winConsoleImplementation"])
+		# Translators: This is the label for a combo box for selecting the
+		# active console implementation in the advanced settings panel.
+		# Choices are automatic, prefer UIA, and force legacy.
+		consoleComboText = _("Windoes C&onsole support:")
```
Comment (leonardder) [inline source/gui/settingsDialogs.py]: Typo: windoes

File: source/gui/settingsDialogs.py [hunk 5]
```diff
@@ -2050,13 +2050,36 @@ def __init__(self, parent):
...
+			#Translators: A choice in a combo box in the advanced settings
+			# panel to have NVDA use UIA in the Windows Console when available.
+			_("Prefer UIA"),
+			#Translators: A choice in a combo box in the advanced settings
+			# panel to have NVDA use its legacy Windoes Console support
+			# in all cases.
+			_("Force legacy")
+		]
+		self.consoleCombo = UIAGroup.addLabeledControl(consoleComboText, wx.Choice, choices=consoleChoices)
+		consoleValsToIndices = {
```
Comment (leonardder) [inline source/gui/settingsDialogs.py]: Please make this a list. Rather than using dict[key], you can use list.index(key)

File: source/gui/settingsDialogs.py [hunk 6]
```diff
@@ -2050,13 +2050,36 @@ def __init__(self, parent):
...
+			"auto": 0,
+			"UIA": 1,
+			"legacy": 2
+		}
+		curChoice = consoleValsToIndices[
+			config.conf['UIA']['winConsoleImplementation']
+		]
+		self.consoleCombo.SetSelection(curChoice)
```
Comment (leonardder) [inline source/gui/settingsDialogs.py]: Isn't `auto` supposed to be the default here?

File: source/gui/settingsDialogs.py [hunk 7]
```diff
@@ -2171,10 +2194,15 @@ def onSave(self):
 		log.debug("Saving advanced config")
 		config.conf["development"]["enableScratchpadDir"]=self.scratchpadCheckBox.IsChecked()
 		config.conf["UIA"]["useInMSWordWhenAvailable"]=self.UIAInMSWordCheckBox.IsChecked()
-		if self.ConsoleUIACheckBox.IsChecked():
-			config.conf['UIA']['winConsoleImplementation'] = "UIA"
-		else:
-			config.conf['UIA']['winConsoleImplementation'] = "auto"
+		consoleChoice = self.consoleCombo.GetSelection()
+		consoleIndicesToVals = {
```
Comment (leonardder) [inline source/gui/settingsDialogs.py]: Please remove this in favour of the list as suggested above. You can save that list as an attribute on the panel. There are numerous examples within the NVDA code base how this is handled, e.g. focus context presentation.

File: user_docs/en/userGuide.t2t [hunk 8]
```diff
@@ -1671,11 +1671,16 @@ This includes in Microsoft Word itself, and also the Microsoft Outlook message v
...
+==== Windows Console support ====[AdvancedSettingsConsoleUIA]
+This option selects how NVDA interacts with the Windows Console used by command prompt, PowerShell, and the Windows Subsystem for Linux. In Windows 10 version 1709, Microsoft [added support for its UI Automation API to the console https://devblogs.microsoft.com/commandline/whats-new-in-windows-console-in-windows-10-fall-creators-update/], bringing vastly improved performance and stability for screen readers that support it. In situations where UI Automation is unavailable or known to result in an inferior user experience, NVDA's legacy console support is available as a fallback.
```
Comment (leonardder) [inline user_docs/en/userGuide.t2t]: Please start every new sentence on a new line.

File: source/gui/settingsDialogs.py [hunk 9]
```diff
@@ -2050,13 +2050,36 @@ def __init__(self, parent):
...
+		curChoice = consoleValsToIndices[
+			config.conf['UIA']['winConsoleImplementation']
+		]
+		self.consoleCombo.SetSelection(curChoice)
+		self.consoleCombo.defaultValue = consoleValsToIndices[
+			self._getDefaultValue(["UIA", "winConsoleImplementation"])
+		]
 
 		# Translators: This is the label for a checkbox in the
 		#  Advanced settings panel.
```
Comment (leonardder) [inline source/gui/settingsDialogs.py]: Could you make it so that this checkbox is hidden when the former combo box is set to legacy? See for an example how the open scratchpad dir button is activated when the checkbox is ticked.

**Review by LeonarddeR — DISMISSED** [review]
Good to see that this is getting in to shape.

File: source/UIAUtils.py (hunk 2 above)
Comment (codeofdusk) [reply source/UIAUtils.py]: @michaelDCurran How confident are you that this will be fixed in 2003? We're at least 8.5 months from release...

File: source/gui/settingsDialogs.py (hunk 6 above)
Comment (codeofdusk) [reply source/gui/settingsDialogs.py]: Doesn't this just get it from the config (which defaults to auto in any case)?

**Comment by codeofdusk** [comment]
Once #9773 is merged, I think this PR is ready for master (pending code review, of course).

**Comment by lukaszgo1** [comment]
It might be worth to have some discussion about when issue is considered closed. In my view one of the following should happen:
1. The issue is fixed - doesn't apply to any of the bugs listed above.
2. The issue is impossible to fix - also not applicable because whine proper fixes aren't easy in these cases it is doable.
3. The product which is affected by the issue is either EOL or NVDA developers decided to no longer support it. It is important to point out that both Windows 7 and Windows 8.1 would be supported by Microsoft until ad least 2023, and LTSB version  of Windows 10 1607 on which Server 2016 is based until 2026.

In short I am against closing issues with the previous console implementations until they are properly fixed. The same has been done when UIA for Word was introduced - no one closed any bug report for COM implementation.

File: source/gui/settingsDialogs.py (hunk 6 above)
Comment (leonardder) [reply source/gui/settingsDialogs.py]: Looks like this comment ended up at the wrong place.

**Comment by feerrenrut** [comment]
@lukaszgo1 Could you be specific about which issue you don't think should be closed to make sure we are on the same page?

File: source/gui/settingsDialogs.py [hunk 10]
```diff
@@ -2050,20 +2051,51 @@ def __init__(self, parent):
...
+			# panel to have NVDA determine its Windows Console implementation
+			# automatically.
+			_("Automatic"),
+			#Translators: A choice in a combo box in the advanced settings
+			# panel to have NVDA use UIA in the Windows Console when available.
+			_("Prefer UIA"),
+			#Translators: A choice in a combo box in the advanced settings
+			# panel to have NVDA use its legacy Windoes Console support
+			# in all cases.
+			_("Force legacy")
```
Comment (feerrenrut) [inline source/gui/settingsDialogs.py]: To prevent confusion when/if we become unable to support legacy in a future windows build, I think it might be more user friendly to name this "Prefer lagacy"

File: user_docs/en/userGuide.t2t [hunk 11]
```diff
@@ -1671,11 +1671,20 @@ This includes in Microsoft Word itself, and also the Microsoft Outlook message v
...
 We still do not recommend that  the majority of users turn this on by default, though we do welcome users of Office 2016/365 to test this feature and provide feedback.
 
-==== Use UI Automation to access the Windows Console when available ====[AdvancedSettingsConsoleUIA]
-When this option is enabled, NVDA will use a new, work in progress version of its support for Windows Console which takes advantage of [accessibility improvements made by Microsoft https://devblogs.microsoft.com/commandline/whats-new-in-windows-console-in-windows-10-fall-creators-update/]. This feature is highly experimental and is still incomplete, so its use is not yet recommended. However, once completed, it is anticipated that this new support will become the default, improving NVDA's performance and stability in Windows command consoles.
+==== Windows Console support ====[AdvancedSettingsConsoleUIA]
+This option selects how NVDA interacts with the Windows Console used by command prompt, PowerShell, and the Windows Subsystem for Linux.
```
Comment (feerrenrut) [inline user_docs/en/userGuide.t2t]: It might be worth explicitly differentiating from "Windows terminal (currently in preview)". Since Windows Terminal runs cmd, powershell and WSL, users could easily get confused here.

File: user_docs/en/userGuide.t2t [hunk 12]
```diff
@@ -1671,11 +1671,20 @@ This includes in Microsoft Word itself, and also the Microsoft Outlook message v
...
+==== Speak passwords in Windows Console ====[AdvancedSettingsWinConsoleSpeakPasswords]
+This setting controls whether characters are spoken by [speak typed characters #KeyboardSettingsSpeakTypedCharacters] or [speak typed words #KeyboardSettingsSpeakTypedWords] in situations where the screen does not update (such as password entry) in the Windows Console with UI automation support enabled.
+For security purposes, this setting should be left disabled.
+However, you may wish to enable it if you experience performance issues or instability with typed character and/or word reporting in consoles, or work in secure environments and prefer password announcement.
```
Comment (feerrenrut) [inline user_docs/en/userGuide.t2t]: I would change this to "or work in trusted environments" rather than in secure environments.

**Comment by lukaszgo1** [comment]
@feerrenrut well I meant all tickets listed in the PR description by @codeofdusk namely: #513, #673, #1682 and #6291.

File: source/UIAUtils.py [hunk 13]
```diff
@@ -203,3 +205,17 @@ def getValue(self,ID,ignoreMixedValues=False):
...
+	"""
+	if not setting:
+		setting = config.conf['UIA']['winConsoleImplementation']
+	if setting == "legacy":
+		return False
+	elif setting == "UIA":
+		return True
+	# #7497: Windows 10 Fall Creators Update has an incomplete UIA implementation for console windows, therefore for now we should ignore it.
+	# It does not implement caret/selection, and probably has no new text events.
+	return isWin10(1803)
```
Comment (feerrenrut) [inline source/UIAUtils.py]: The docs for `isWin10` say `@note: this function will always return False for source copies of NVDA due to a Python bug.`

Is this true? if not, please update the docs for the `isWin10` function.

Comment (leonardder) [reply source/UIAUtils.py]: It is True for python 2 and false for python 3.

Comment (codeofdusk) [reply source/UIAUtils.py]: Yes, this is definitely at least true for Python 2.

Comment (feerrenrut) [reply source/UIAUtils.py]: Ah, I think we are all on the same page, but I just realised my question "is it true" is quite open to interpretation. I was asking if the docs are correct. Please correct me if I'm mistaken with the following:
- Python 2: running from source causes the `isWin10` function to behave incorrectly, it always returns False.
- Python 3: the bug is fixed, `isWin10` function returns True when running from source on a Windows 10 machine.

If this makes it into Python 2 code (which is currently not the plan), we would need to look at this carefully to see if there is any way we can improve the situation, given how prevalent using the console is for people working from source copies. For now, no need to do anything, Bill could you add this to your checklist for the end of your project?

File: user_docs/en/userGuide.t2t [hunk 14]
```diff
@@ -1672,19 +1672,19 @@ However, There may be some information which is either not exposed, or exposed i
 We still do not recommend that  the majority of users turn this on by default, though we do welcome users of Office 2016/365 to test this feature and provide feedback.
 
 ==== Windows Console support ====[AdvancedSettingsConsoleUIA]
-This option selects how NVDA interacts with the Windows Console used by command prompt, PowerShell, and the Windows Subsystem for Linux.
+This option selects how NVDA interacts with the Windows Console used by command prompt, PowerShell, and the Windows Subsystem for Linux. It does not affect the modern Windows Terminal.
```
Comment (feerrenrut) [inline user_docs/en/userGuide.t2t]: Please add this sentence on a new line.

File: user_docs/en/userGuide.t2t [hunk 15]
```diff
@@ -1672,19 +1672,19 @@ However, There may be some information which is either not exposed, or exposed i
...
 In Windows 10 version 1709, Microsoft [added support for its UI Automation API to the console https://devblogs.microsoft.com/commandline/whats-new-in-windows-console-in-windows-10-fall-creators-update/], bringing vastly improved performance and stability for screen readers that support it.
 In situations where UI Automation is unavailable or known to result in an inferior user experience, NVDA's legacy console support is available as a fallback.
 The Windows Console support combo box has three options:
 - Automatic: Uses UI Automation in consoles on Windows 10 version 1803 and later. This option is recommended and set by default.
 - Prefer UIA: Uses UI Automation in consoles if available, even on Windows versions with incomplete or buggy implementations. While this limited functionality may be useful (and even sufficient for your usage), use of this option is entirely at your own risk and no support for it will be provided.
-- Force legacy: UI Automation in the Windows Console will be completely disabled, so the legacy fallback will always be used.
+- Legacy: UI Automation in the Windows Console will be completely disabled, so the legacy fallback will always be used.
```
Comment (feerrenrut) [inline user_docs/en/userGuide.t2t]: Great!

**Review by feerrenrut — APPROVED** [review]
Just a minor change to put a new sentence on a new line in the user guide, otherwise I'm happy for this to be merged. To ensure that the Python 3 merge goes smoothly, I'll hold off on merging this until we have merged back from threshold.

**Review by feerrenrut — APPROVED** [review]
Thanks @codeofdusk
